synapse/cortex.py

```diff
@@ -914,50 +875,63 @@ def getCoreMods(self):
             ret.append((modname, mod.conf))
         return ret
 
-    def eval(self, text, opts=None, user=None):
-        with self.snap(user=user) as snap:
-            yield from snap.eval(text, opts=opts, user=user)
-
-    def storm(self, text, opts=None, user=None):
-
-        chan = s_queue.Queue()
-
-        # check syntax before executing...
-        # (the query will be cached anyway )
-        self.getStormQuery(text)
-
-        self._runStormThread(text, opts=opts, user=user, chan=chan)
-
-        for mesg in chan:
-            yield mesg
-
-    @s_glob.inpool
-    def _runStormThread(self, text, opts=None, user=None, chan=None):
-
-        count = 0
-        tick = s_common.now()
-
-        try:
-
-            chan.put(('init', {'tick': tick, 'text': text}))
-
-            with self.snap(user=user) as snap:
-
-                snap.link(chan.put)
+    async def eval(self, text, opts=None, user=None):
+        '''
+        Evaluate a storm query and yield Nodes only.
+        '''
+        async with await self.snap(user=user) as snap:
+            async for node in snap.eval(text, opts=opts):
+                yield node
 
-                for pode in snap.iterStormPodes(text, opts=opts, user=user):
-                    chan.put(('node', pode))
-                    count += 1
+    async def storm(self, text, opts=None, user=None):
+        '''
+        Evaluate a storm query and yield result messages.
+        Yields:
+            (Node, Path) tuples
+        '''
+        async with await self.snap(user=user) as snap:
+            async for mesg in snap.storm(text, opts=opts):
+                yield mesg
 
-        except Exception as e:
-            logger.exception('Error during Storm execution.')
-            chan.put(('err', s_common.err(e)))
+    async def streamstorm(self, text, opts=None, user=None):
```

Comment:
There is a subtle change in the order of operations with this implementation which concerns me.  In the `storm` to -> `_runStormThread` call chain, a bad query would die prior to a `init` message being sent over the channel.  Now, that is kind-of caught with the `count` variable and passed along AFTER the `init` message has been sent.  In addition, a generally valid query which never yields a node which has a unhandled exception will die instead of being caught, wrapped in a `err` and passed along.  

We should revert these changes so that we die right away with a bad query; and do our best to ensure that we will always deliver the stream of messages (init, node, err, fini) into the channel once we start streaming them.